hive-metastore/src/main/java/org/apache/iceberg/hive/MetastoreLock.java

```diff
@@ -0,0 +1,533 @@
+                    LOG.warn(
+                        "Interrupted while waiting for lock on table {}.{}",
+                        databaseName,
+                        tableName,
+                        e);
+                  }
+                },
+                TException.class);
+      }
+    } catch (WaitingForLockException e) {
+      timeout = true;
+      duration = System.currentTimeMillis() - start;
+    } catch (TException e) {
+      error = e;
+    } finally {
+      if (!lockInfo.lockState.equals(LockState.ACQUIRED)) {
+        unlock(Optional.of(lockInfo.lockId));
+      }
+    }
+
+    if (lockInfo.lockState.equals(LockState.ACQUIRED)) {
+      return lockInfo.lockId;
+    }
+
+    // timeout and do not have lock acquired
+    if (timeout) {
+      throw new LockException(
+          "Timed out after %s ms waiting for lock on %s.%s", duration, databaseName, tableName);
+    }
+
+    // On thrift error and do not have lock acquired
+    if (error != null) {
+      throw new LockException(
+          error, "Metastore operation failed for %s.%s", databaseName, tableName);
+    }
+
+    // Just for safety. We should not get here.
+    throw new LockException(
+        "Could not acquire the lock on %s.%s, lock request ended in state %s",
+        databaseName, tableName, lockInfo.lockState);
+  }
+
+  /**
+   * Tries to create a lock. If the lock creation fails, and it is possible then retries the lock
+   * creation a few times. If the lock creation is successful then a {@link LockInfo} is returned,
+   * otherwise an appropriate exception is thrown.
+   *
+   * @return The created lock
+   * @throws LockException When we are not able to fill the hostname for lock creation, or there is
+   *     an error during lock creation
+   */
+  @SuppressWarnings("ReverseDnsLookup")
+  private LockInfo tryLock() throws LockException {
+    LockInfo lockInfo = new LockInfo();
+
+    String hostName;
+    try {
+      hostName = InetAddress.getLocalHost().getHostName();
+    } catch (UnknownHostException uhe) {
+      throw new LockException(uhe, "Error generating host name");
+    }
+
+    final LockComponent lockComponent =
+        new LockComponent(LockType.EXCLUSIVE, LockLevel.TABLE, databaseName);
+    lockComponent.setTablename(tableName);
+    final LockRequest lockRequest =
+        new LockRequest(
+            Lists.newArrayList(lockComponent), System.getProperty("user.name"), hostName);
+
+    // Only works in Hive 2 or later.
+    if (HiveVersion.min(HiveVersion.HIVE_2)) {
+      lockRequest.setAgentInfo(agentInfo);
+    }
+
+    AtomicBoolean interrupted = new AtomicBoolean(false);
+    Tasks.foreach(lockRequest)
+        .retry(Integer.MAX_VALUE - 100)
+        .exponentialBackoff(
+            lockCreationMinWaitTime, lockCreationMaxWaitTime, lockCreationTimeout, 2.0)
+        .shouldRetryTest(e -> !interrupted.get() && HiveVersion.min(HiveVersion.HIVE_2))
```

Comment:
No need to keep e instanceof TException here?

Reply:
`HiveTableOperations.doCommit` used to handle `TException`, `InterruptedException`, `UnknownHostException` before the change by converting them to a `RuntimeException`. After the change we throw a `CommitFailedException` which is a `RuntimeException`.

It is definitely some change in the API, but I think the `CommitFailedException` better describes the situation than a random `RuntimeException`.

TBH I would change the remaining 2 places where we throw `RuntimeException`s (wrapping `TException` and `InterruptedException`) from the `loadHmsTable`, but we have specific tests for these, so I decided against it.

We can continue the discussion on #6648